### ISSUE TITLE
storage/concurrency: prep lock-table for 20.1 release

### ISSUE THREAD

**Issue by nvanbenschoten** [cockroachdb/cockroach#44976]
Release 20.1 will contain a new approach to kv-level transaction locking. Instead of storing all locks inline in the MVCC keyspace and finding locks during request evaluation, locks will now be maintained in a separate lock-table. While replicated locks (intents) will not be pulled into a persistent lock-table keyspace during this release, the lock-table structure will track these locks, as well as unreplicated (best-effort) locks which enable SELECT FOR UPDATE (see #40205). Additionally, the lock-table structure will lead to significant improvements in transaction queuing. The lock-table's lock wait-queues will replace the current `contentionQueue`, leading to improved fairness between transactions and reduced latency now that the lock wait-queues react directly to intent resolution. For more on this change, see https://github.com/cockroachdb/cockroach/issues/41720.

The rest of this issue enumerates the remaining work items for the lock-table and `storage/concurrency` package that should get into this release. It does not include any work items that are specific to SELECT FOR UPDATE's use of the `storage/concurrency` package.

### Must Have
- [x] pull Store-level concurrency retry loop under Replica: @nvanbenschoten #43138
- [x] stop modifying requests and returning responses from TxnWaitQueue: @nvanbenschoten #43383
- [x] prep LocalResult struct to carry LockUpdate and LockAcquisition events to feed into lock-table: @nvanbenschoten #43703
- [x] define lock Strength and Durability modes: @nvanbenschoten #44054
- [x] create lock-table and lock wait-queue structures: @sumeerbhola #43740
- [x] generate type-safe specializations of interval B-tree: @nvanbenschoten #43850
- [x] define concurrency control interfaces: @nvanbenschoten #44787
- [x] randomized test for lockTable with concurrency: @sumeerbhola #44791
- [x] implement lockTableWaiter interface: @nvanbenschoten #44885
- [x] micro-benchmark lockTable: @sumeerbhola #44964
- [x] implement concurrency.Manager interface: @nvanbenschoten #45062
- [x] don't push reservation holders, only lock holders: @nvanbenschoten #45420
- [x] hook concurrency.Manager structure into storage package: @nvanbenschoten #45482
- [x] push reservation holders to detect deadlocks: @nvanbenschoten #45567
- [x] replace btree implementation in lockTable with specialized copy-on-write btree: @sumeerbhola #45276 
- [ ] place memory limits on lockTable and degrade gracefully when limits are hit
- [x] add testing for lock upgrades (i.e. with epoch bumps and ignored seq nums): @sumeerbhola #45769
- [x] support non-transactional writes in lock-table: @sumeerbhola #44975
- [x] allow ReadOnly and ReadWrite spans passed to lockTable to overlap: @sumeerbhola #45124 
- [x] make `waitElsewhere` state be used only for a replicated lock that is held: @sumeerbhola #45040 #45080 
- [x] change write timestamp declared by ResolveIntent{Range} requests to txn.MinTimestamp: @nvanbenschoten #45232
- [x]  update waitingState.held on each waiter when locks are acquired and released: @sumeerbhola #45040 
- [x] eliminate the doneWaiting slice that is used as a return value for many lockState methods: @sumeerbhola #45040
- [x] clear queued waiters when part of same txn as discovered lock: @nvanbenschoten #45601 
- [x] optimize memory allocations in lockTable: @nvanbenschoten #45633
- [x] stabilize `kv/contention/nodes=4`: @nvanbenschoten 
- [x] benchmark cockroach with and without concurrency.Manager integration using `kv`, `sysbench`, and `tpcc` to detect regression: @nvanbenschoten 
- [x] disable lock-table when not leaseholder: @nvanbenschoten #45830

### Nice To Have
- [ ] use txn priorities when ordering requests to avoid higher priority request waiting on a lower priority reservation holder @sumeerbhola
- [ ] benchmark comparison between lazy waiting (current) and eager waiting in contended workloads @sumeerbhola  
- [ ] re-introduce invariant that reservation holder is always the request that acquires locks
- [x] distinguish in `lockTableGuardImpl` between waiting on lock holder and waiting on queued requests #44885 
- [x] add visibility into state of lockTable through metrics and/or debug pages: https://github.com/cockroachdb/cockroach/pull/67350
- [ ] simplify state transitions in `acquireLock` by not removing from lock holder's `lockTableGuardImpl.mu.locks`
- [ ] eliminate waiting details except when waiting on lock holder, and do related code simplification @sumeerbhola 

_Feel free to modify. Please check off when changes merge._

**Comment by nvanbenschoten** [comment]
Closing. This issue served us well, but it is now stale.
